### ISSUE TITLE
MultiThreadedEventLoopGroup: allow users to specify the thread name pattern

### ISSUE THREAD

**Issue by weissi** [apple/swift-nio#2473]
Right now, MTELG's thread name pattern is always `NIO-ELT-x-#y`. The `NIO-ELT-` part should be configurable by the user. And `NIOSingletons`'s ELG's (once https://github.com/apple/swift-nio/pull/2471 is in) should set that to something useful.

**Comment by natikgadzhi** [comment]
So the whole code infrastructure for this is already implemented in #2471: https://github.com/apple/swift-nio/pull/2471/files#diff-4b1291d44c519e5e77e9d2d3f165562ca65b739df1286327c5a0b0e1b5e98ff1R199

It's the API design that is the challenging part?

I'm happy to poke you (and @glbrntt) with a pull request with a draft — that tends to spawn pretty fun discussions, it seems 🙃 (At least fun for me)

> And NIOSingletons's ELG's (once https://github.com/apple/swift-nio/pull/2471 is in) should set that to something useful.

As in, automatically default to something other than `"NIO-ELT-"`? Hmmm. Any chance we could grab the current swift module name the code is compiled in when `NIOSingletons` is invoked? I mean, we could slice the path of `#file` 👀, right? Then we could do something like `NIO-MODULE-%idx%`.

**Comment by Lukasa** [comment]
I do not believe it is possible to get the module name of the invoking module.

**Comment by weissi** [comment]
> I do not believe it is possible to get the module name of the invoking module.

`#fileID` is _guaranteed_ to be `<module name>/<file base name>.swift` so it is _possible_ but requires parsing.

**Comment by weissi** [comment]
Draft implementation in #2943
